Fix comparison of bit shifts (unchecked overflow) to arithmetic (checked overflow) (#11388)
ethereum/solidity: docs/types/value-types.rst

Shifts
^^^^^^

> The result of a shift operation has the type of the left operand, truncating the result to match the type.

Reviewer: Ah, this seems to be missing. I'll fix this.

> This means that contrary to checked arithmetic, if the result of a bit shift operation will not fit its type, the transaction will *not* revert.

Reviewer 2: What is "This" referring to here?

> The result of a shift operation has the type of the left operand, truncating the result to match the type. This means an "overflow" will not be "checked".

Reviewer 3: Truncating the result pretty much means that there will be no overflow checks - so in that sense I would actually say that this *was* documented. But it for sure doesn't hurt to be absolutely clear and explicit.
To me the quoting seems a bit off, though. What about
```suggestion
This means that contrary to checked arithmetic, if the result of a bit shift operation will not fit its type, the transaction will *not* revert.
```
or something like that?

Author: Good wording, thank you

Reviewer 3: Sorry, apparently I mistakenly erased the first sentence in my suggestion, so now it vanished - we'll want to keep that.

Right operand must be unsigned type. Trying to shift by signed type will produce a compilation error.

> - For positive and negative ``x`` values, ``x << y`` is equivalent to ``x * 2**y`` run in unchecked mode.

Reviewer: We do have this documented in the section about `checked` and `unchecked`: https://docs.soliditylang.org/en/v0.8.4/control-structures.html?highlight=bitwise%20operators%20do%20not#checked-or-unchecked-arithmetic

> Bitwise operators do not perform overflow or underflow checks. This is particularly visible when using bitwise shifts (<<, >>, <<=, >>=) in place of integer division and multiplication by a power of 2. For example type(uint256).max << 3 does not revert even though type(uint256).max * 8 would.

----

Anyway, good to have it mentioned here as well.

> - For positive and negative ``x`` values, ``x << y`` is equivalent to ``x * 2**y`` run in unchecked mode.

Reviewer 2: Nitpick, but that's (for the better or worse) it is referred to in other parts of the documentation:
```suggestion
- For positive and negative ``x`` values, ``x << y`` is equivalent to ``x * 2**y`` run in "unchecked" mode.
```

> - For positive and negative ``x`` values, ``x << y`` is equivalent to ``x * 2**y`` run in unchecked mode.

Reviewer 3: This is indeed a good spot, this definitely has to be changed and/or clarified. Strictly speaking, this applies to the other cases as well, though, the exponentiation can just as well overflow in the right shift case and things still won't revert, whereas in checked arithmetic the supposedly "equivalent" expression would.

^^^^^^

The result of a shift operation has the type of the left operand, truncating the result to match the type.
The right operand must be of unsigned type, trying to shift by an signed type will produce a compilation error.

> - For positive and negative ``x`` values, ``x << y`` is equivalent to ``x * 2**y`` run in "unchecked" mode.

Reviewer 3: Strictly speaking ``run in "unchecked" mode`` would need to be added to the ones below as well... ``x >> 300`` won't panic...

Reviewer 4: maybe we should rephrase this altogether then and turn it into a mathematical expression instead of solidity code?

^^^^^^

The result of a shift operation has the type of the left operand, truncating the result to match the type.
The right operand must be of unsigned type, trying to shift by an signed type will produce a compilation error.

Shifts can be "simulated" using multiplication by powers of two in the following way. Note that the truncation
to the type of the left operand is also performed at the end, but not mentioned explicitly.

> - For positive and negative ``x`` values, ``x << y`` is equivalent to the mathematical expression ``x * 2**y``, rounded towards zero.

Reviewer: Why is rounded towards zero used here?

> - For positive ``x`` values,  ``x >> y`` is equivalent to the mathematical expression ``x / 2**y``, rounded towards zero.

Reviewer: Why is rounded towards zero important here? The value is always positive.

Maybe it's better to say round down or change the expression to `floor(x / 2**y)` just like in EIP-145.

Reviewer 4: I think "towards zero" is clearer than "down", even if the latter might be more commonly used.

- For negative ``x`` values, ``x >> y`` is equivalent to the mathematical expression ``(x + 1) / 2**y - 1`` (which is the same as dividing ``x`` by ``2**y`` while rounding down towards negative infinity).

.. warning::
    Before version ``0.5.0`` a right shift ``x >> y`` for negative ``x`` was equivalent to
    the mathematical expression ``x / 2**y`` rounded towards zero,
    i.e., right shifts used rounding up (towards zero) instead of rounding down (towards negative infinity).

.. note::

>     Overflows checks are never performed for shift operations as they are done for arithmetic operations.

Reviewer: ```suggestion
    Overflow checks are never performed for shift operations as they are done for arithmetic operations.
```
